server/api/rounds.py

```diff
@@ -448,23 +478,34 @@ def draw_sample(election: Election, round: Round, sample_sizes: Dict[str, int]):
     )
 
     # Create RoundContest objects to include the contests in this round
-    for contest in contests_that_havent_met_risk_limit:
-        round_contest = RoundContest(
-            round_id=round.id,
-            contest_id=contest.id,
+    round.round_contests = [
+        RoundContest(
+            round=round,
+            contest=contest,
             # Store the sample size we use for each contest so we can report on
             # it later. Opportunistic contests don't have a sample size, so we
             # store None.
             sample_size=sample_sizes.get(contest.id, None),
         )
-        db_session.add(round_contest)
+        for contest in contests_that_havent_met_risk_limit
+    ]
 
     contests_to_sample = [
         contest
         for contest in contests_that_havent_met_risk_limit
         if contest.is_targeted
     ]
 
+    # Special case: if we are sampling all ballots, we don't need to actually
+    # draw a sample. Instead, we force an offline audit.
+    if sampled_all_ballots(round, election):
+        if len(contests_to_sample) > 1:
+            raise Conflict(
+                "Cannot sample all ballots when there are multiple targeted contests."
+            )
+        election.online = False
```

Comment:
no `db_session.add(election)` ?

Is this code supposed to happen on every round or just the first? I assume it can only happen on round 1, given that we're looking at everything but... just checking.

Reply:
Similar to above, I think when you've loaded a SQLAlchemy object, it's loaded into the session, so you don't need to re-add it to the session when you modify it. I think you only need to add new objects you're creating to the session.

I agree that there can only be one round in this case. But it wouldn't hurt for this particular line to run every round.